File: app/src/main/kotlin/com/spacex/ui/rocketDetails/launches/RocketLaunchUIMDecorator.java
package com.spacex.ui.rocketDetails.launches;

import static com.spacex.ui.rocketDetails.launches.LongExtensionFunctionsKt.getEitherDayAndMonthWithOptionallyWeekdayAndYearOrRelativeTimeSpan;

import android.content.Context;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import com.spacex.ui.R;

public class RocketLaunchUIMDecorator {

  private final Context context;
  private RocketLaunchUIM rocketLaunchUIM;

  RocketLaunchUIMDecorator(@NonNull final Context context) {
    this.context = context;
  }

  @NonNull
  public String getVideoKey() {
    final String videoUrl = rocketLaunchUIM.getVideoUrl();
    return videoUrl.substring(videoUrl.indexOf("=") + 1, videoUrl.length());
  }

  @NonNull
  public String getMissionName() {
    return rocketLaunchUIM.getMissionName();
  }

  @NonNull
  public String getMissionPatchLink() {
    return rocketLaunchUIM.getMissionPatchUrl();
  }

  @NonNull
  public SpannableString getMissionDetails() {
    final String missionDetails =
        context.getString(
            R.string.rocket_launch_details_format,
            getEitherDayAndMonthWithOptionallyWeekdayAndYearOrRelativeTimeSpan(
                rocketLaunchUIM.getDate()),
            getSuccessOrFailure(rocketLaunchUIM.getWasSuccess()));
    final SpannableString missionDetailsSpan = new SpannableString(missionDetails);
    missionDetailsSpan.setSpan(
        new ForegroundColorSpan(
            ContextCompat.getColor(context, rocketLaunchUIM.getSuccessLabelColorResId())),
        missionDetails.indexOf("•") + 1,
        missionDetails.length(),
        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
    return missionDetailsSpan;
  }

  RocketLaunchUIMDecorator onBind(@NonNull final RocketLaunchUIM rocketLaunchUIM) {
    this.rocketLaunchUIM = rocketLaunchUIM;
    return this;
  }

  private String getSuccessOrFailure(final boolean wasSuccessful) {
    return context.getString(wasSuccessful ? R.string.successful : R.string.failure);
  }
}

File: app/src/main/kotlin/com/spacex/ui/rocketDetails/launches/RocketLaunchUIMDecorator.kt
package com.spacex.ui.rocketDetails.launches

import android.content.Context
import android.text.Spannable
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import androidx.core.content.ContextCompat
import com.spacex.ui.R

internal class RocketLaunchUIMDecorator(private val context: Context) {
    private var rocketLaunchUIM: RocketLaunchUIM? = null

    fun onBind(rocketLaunchUIM: RocketLaunchUIM): RocketLaunchUIMDecorator {
        this.rocketLaunchUIM = rocketLaunchUIM
        return this
    }

    val videoKey: String
        get() {
            val videoUrl = rocketLaunchUIM!!.videoUrl
            return videoUrl.substring(videoUrl.indexOf("=") + 1, videoUrl.length)
        }

    val missionName: String
        get() = rocketLaunchUIM!!.missionName

    val missionPatchLink: String
        get() = rocketLaunchUIM!!.missionPatchUrl

    val missionDetails: SpannableString
        get() {
            val missionDetails = context.getString(
                    R.string.rocket_launch_details_format,
                    rocketLaunchUIM!!.date.getEitherDayAndMonthWithOptionallyWeekdayAndYearOrRelativeTimeSpan(),
                    getSuccessOrFailure(rocketLaunchUIM!!.wasSuccess))
            val missionDetailsSpan = SpannableString(missionDetails)
            missionDetailsSpan.setSpan(
                    ForegroundColorSpan(
                            ContextCompat.getColor(context, rocketLaunchUIM!!.getSuccessLabelColorResId())),
                    missionDetails.indexOf("•") + 1,
                    missionDetails.length,
                    Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
            return missionDetailsSpan
        }

    private fun getSuccessOrFailure(wasSuccessful: Boolean): String {
        return context.getString(if (wasSuccessful) R.string.successful else R.string.failure)
    }
}
